### DISCUSSION TITLE
feature request: prefer SHELL from `launchctl getenv SHELL` or `systemctl show-environment` when available

### DISCUSSION THREAD

**Discussion by jokeyrhyme** [ghostty-org/ghostty#3598] [Feature Requests, Ideas] [answered]
I'm on a company MacBook, which means I can't modify anything outside of my home directory

This means that I cannot modify /etc/shells to register an alternative shell like `fish` or `nu` (nushell)

So, my current workaround is to run `launchctl setenv SHELL /opt/homebrew/bin/nu` before running any other apps, but this only registers the new environment variable value in the user service manager (e.g. launchd or systemd) so it's currently a little hit-and-miss as to which apps read from there or not

I know I can also put `command = "/opt/homebrew/bin/nu"` in my Ghostty settings, however, I use the same settings on my Linux machines, so this makes my settings no longer portable between Linux and macOS

I'd understand completely if the project decided this was out-of-scope or not worth the complexity, but it'd be terrific if Ghostty could check the user service manager (launchd or systemd) if present for SHELL first, then fallback to SHELL in the process's current environment variables (which is currently does)

Cheers! <3

**Comment by mitchellh** [comment]
Thanks for the writeup. Is there precedent for other software or terminals doing this?

**Reply by jokeyrhyme** [reply]
It's not so much specific terminals/apps that do this, but every new process launched as part of the `systemd` unit tree will have all of the latest variable values that were set in the service manager, which in practice applies to every new process in a desktop environment

Unfortunately, `launchd` on macOS seems abandoned by its maintainer, so even though it could very easily offer the same experience, this particular aspect is up to each app developer on that platform :S

I've been able to use `launchctl setenv ...` to preserve values between shell sessions (e.g. SSH_AUTH_SOCK, etc) so that they don't need to be reinitialised between terminal tabs/windows, but this only works because I've configured by shell to `launchctl getenv ...` during initialisation, but this is too late to apply to the first shell process launched by Ghostty

I think really I've got an X/Y problem happening here, as the root problem I'd like to solve is "how do I have the one static `git`-versioned Ghostty configuration work between multiple machines with various inconvenient differences?"

**Comment by jokeyrhyme** [comment]
> I think really I've got an X/Y problem happening here, as the root problem I'd like to solve is "how do I have the one static `git`-versioned Ghostty configuration work between multiple machines with various inconvenient differences?"

Another way to solve this could be to be able to set `command = ...` to multiple executables, so that Ghostty checks to see if they exist and runs the first one that does

**Comment by jokeyrhyme** [comment]
> I think really I've got an X/Y problem happening here, as the root problem I'd like to solve is "how do I have the one static `git`-versioned Ghostty configuration work between multiple machines with various inconvenient differences?"

Another way to solve this could be to optionally have Ghostty run a login shell in order to harvest user-customised values

The GNOME and COSMIC desktops on Linux do this just prior to bringing up the graphical environment: https://github.com/pop-os/cosmic-session/blob/0a3b9f5376a089b0aeb6432a663b0adc17c48746/data/start-cosmic#L21-L29

However, this is extra work that would have to be done during Ghostty initialisation, so I could see why this wouldn't be desirable by default or even at all 🤷

**Reply by 00-kat** [reply]
FYI, Ghostty does currently run a login shell on macOS: https://github.com/ghostty-org/ghostty/blob/6cbd69da7839260508466f9dfb2bc0c0fbb43991/src/termio/Exec.zig#L947-L99.

**Reply by jokeyrhyme** [reply]
@00-kat 

interesting ...

1. I've run `launchctl setenv SHELL /opt/homebrew/bin/nu`

2. I've put the following in my `~/.zshenv`:

```sh
SHELL="$(/bin/launchctl getenv SHELL)"
export SHELL
```

3. I've completely quit Ghostty, removed `command = "/opt/homebrew/bin/nu"` from its settings, and restarted it

4. `env` still shows `SHELL=/bin/zsh` and this is what is running, not `/opt/homebrew/bin/nu`

When you say Ghostty runs a login shell, do you mean it does this before running the final shell that will be displayed to the user, to prepare for running that shell? Or do you mean it runs that shell as a login shell?

Apologies, I wasn't clear in my suggestion here, what I meant was:

1. Ghostty runs the SHELL or /etc/passwd (or equivalent) shell as a login shell and harvests the environment variables it sets (`start-cosmic` does this via a round-trip through `exec`)

2. Ghostty now uses the possibly-updated SHELL value to present to the user

**Comment by jokeyrhyme** [comment, marked as answer]
> I think really I've got an X/Y problem happening here, as the root problem I'd like to solve is "how do I have the one static git-versioned Ghostty configuration work between multiple machines with various inconvenient differences?"

Okay, there's already a possible workaround here:

https://ghostty.org/docs/config#splitting-into-multiple-files

I can put all my common settings in a `git`-versioned file that is shared between all of my machines, and include a `config-file = ?machine-specific` entry in those settings

And then populate that "machine-specific" file with whatever doesn't make sense to share between machines
